pageserver/src/tenant.rs

```diff
@@ -350,9 +361,604 @@ impl Drop for TimelineUninitMark {
+                    tenant_dir.display(), marker_file.display());
+            }
+        } else {
+            crashsafe::create_dir_all(&tenant_dir).context("create tenant directory")?;
+            fs::File::create(&marker_file).context("create tenant attaching marker file")?;
+            crashsafe::fsync_file_and_parent(&marker_file)
+                .context("fsync tenant attaching marker file and parent")?;
+        }
+        debug_assert!(tenant_dir.is_dir());
+        debug_assert!(marker_file.is_file());
+
+        // Get list of remote timelines
+        // download index files for every tenant timeline
+        info!("listing remote timelines");
+
+        let remote_storage = self
+            .remote_storage
+            .as_ref()
+            .ok_or_else(|| anyhow::anyhow!("cannot attach without remote storage"))?;
+
+        let remote_timelines =
+            list_remote_timelines(remote_storage, self.conf, self.tenant_id).await?;
+
+        info!("found {} timelines", remote_timelines.len());
+
+        let mut timeline_ancestors: HashMap<TimelineId, TimelineMetadata> = HashMap::new();
+        let mut index_parts: HashMap<TimelineId, IndexPart> = HashMap::new();
+        for (timeline_id, index_part) in remote_timelines {
+            let remote_metadata = index_part.parse_metadata().with_context(|| {
+                format!(
+                    "Failed to parse metadata file from remote storage for tenant {} timeline {}",
+                    self.tenant_id, timeline_id
+                )
+            })?;
+            timeline_ancestors.insert(timeline_id, remote_metadata);
+            index_parts.insert(timeline_id, index_part);
+        }
+
+        // For every timeline, download the metadata file, scan the local directory,
+        // and build a layer map that contains an entry for each remote and local
+        // layer file.
+        let sorted_timelines = tree_sort_timelines(timeline_ancestors)?;
+        for (timeline_id, remote_metadata) in sorted_timelines {
+            // TODO again handle early failure
+            self.load_remote_timeline(
+                timeline_id,
+                index_parts.remove(&timeline_id).unwrap(),
+                remote_metadata,
+                remote_storage.clone(),
+            )
+            .await
+            .with_context(|| {
+                format!(
+                    "failed to load remote timeline {} for tenant {}",
+                    timeline_id, self.tenant_id
+                )
+            })?;
+        }
+
+        std::fs::remove_file(&marker_file)
+            .with_context(|| format!("unlink attach marker file {}", marker_file.display()))?;
+        crashsafe::fsync(marker_file.parent().expect("marker file has parent dir"))
+            .context("fsync tenant directory after unlinking attach marker file")?;
+
+        // FIXME: Check if the state has changed to Stopping while we were downloading stuff
```

Comment:
When can that happen? Can't spot it.

Anyways, there is another race, with Tenant::shutdown.
Can we fix that somehow or do we leave it as is?

Either way, we should update this FIXME to clarify things.

Reply:
Hmm.

Comment says Stopping, but the state is called Paused in this PR. I actually think Stopping is a much better name, but whatever.

If you attach a tenant, and detach it again just before it reaches this FIXME comment, what happens? I wrote a test case for that: https://github.com/neondatabase/neon/compare/dkr/on-demand-split/per-tenant-remote-sync...test-detach-before-attach-completes?expand=1

`detach_tenant` will set the Tenant to Paused state, but then the attach task will continue and call `activate`. There is a period where the Tenant is in Active state, but has already been removed from the global hash table. `detach_tenant` calls `shutdown_tasks`, which waits for the attach-task to finish, before it removes the underlying data directory, so nothing very serious happens. However, the GC and compaction tasks are spawned only after the `shutdown_tasks` has been called, so the GC and compaction loops keep running and failing:
```
2022-11-23T21:47:08.556124Z  INFO POST /v1/tenant/f34dde035dab1d81af5eb9a41df75d4d/detach 200 OK
2022-11-23T21:47:10.557231Z ERROR compaction_loop{tenant_id=f34dde035dab1d81af5eb9a41df75d4d}: Failed to get a tenant f34dde035dab1d81af5eb9a41df75d4d: Tenant f34dde035dab1d81af5eb9a41df75d4d not found in the local state
2022-11-23T21:47:10.557271Z ERROR gc_loop{tenant_id=f34dde035dab1d81af5eb9a41df75d4d}: Failed to get a tenant f34dde035dab1d81af5eb9a41df75d4d: Tenant f34dde035dab1d81af5eb9a41df75d4d not found in the local state
2022-11-23T21:47:12.558667Z ERROR gc_loop{tenant_id=f34dde035dab1d81af5eb9a41df75d4d}: Failed to get a tenant f34dde035dab1d81af5eb9a41df75d4d: Tenant f34dde035dab1d81af5eb9a41df75d4d not found in the local state
2022-11-23T21:47:12.558724Z ERROR compaction_loop{tenant_id=f34dde035dab1d81af5eb9a41df75d4d}: Failed to get a tenant f34dde035dab1d81af5eb9a41df75d4d: Tenant f34dde035dab1d81af5eb9a41df75d4d not found in the local state
2022-11-23T21:47:14.560288Z ERROR compaction_loop{tenant_id=f34dde035dab1d81af5eb9a41df75d4d}: Failed to get a tenant f34dde035dab1d81af5eb9a41df75d4d: Tenant f34dde035dab1d81af5eb9a41df75d4d not found in the local state
2022-11-23T21:47:14.560339Z ERROR gc_loop{tenant_id=f34dde035dab1d81af5eb9a41df75d4d}: Failed to get a tenant f34dde035dab1d81af5eb9a41df75d4d: Tenant f34dde035dab1d81af5eb9a41df75d4d not found in the local state
2022-11-23T21:47:16.562383Z ERROR gc_loop{tenant_id=f34dde035dab1d81af5eb9a41df75d4d}: Failed to get a tenant f34dde035dab1d81af5eb9a41df75d4d: Tenant f34dde035dab1d81af5eb9a41df75d4d not found in the local state
2022-11-23T21:47:16.562454Z ERROR compaction_loop{tenant_id=f34dde035dab1d81af5eb9a41df75d4d}: Failed to get a tenant f34dde035dab1d81af5eb9a41df75d4d: Tenant f34dde035dab1d81af5eb9a41df75d4d not found in the local state
2022-11-23T21:47:18.563936Z ERROR compaction_loop{tenant_id=f34dde035dab1d81af5eb9a41df75d4d}: Failed to get a tenant f34dde035dab1d81af5eb9a41df75d4d: Tenant f34dde035dab1d81af5eb9a41df75d4d not found in the local state
2022-11-23T21:47:18.563996Z ERROR gc_loop{tenant_id=f34dde035dab1d81af5eb9a41df75d4d}: Failed to get a tenant f34dde035dab1d81af5eb9a41df75d4d: Tenant f34dde035dab1d81af5eb9a41df75d4d not found in the local state
```

If you then re-attach the tenant again, and the leftover GC and compaction loops from the previous iteration are still running, things could get really confusing. I couldn't actually reproduce any serious failures with that, but I didn't try very hard.

We need to tighten this up. I'm not sure it needs to be done as part of this PR, though, we have serious race conditions around detach/attach in main too. I'm not sure this is any worse, although that's a very low bar.

It would be pretty simple to check that the tenant is not in Paused state, and only proceed with activation if it's not. Let's do that. There's probably a lot more we could and should do, but that can be left to a followup PR.

Reply:
Opened https://github.com/neondatabase/neon/pull/2910 to address this.